test/jdk/java/util/IdentityHashMap/DefaultRemoveReplace.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (c) 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.util.IdentityHashMap;
+import java.util.Objects;
+
+/*
+ * @test
+ * @bug 8178355
+ * @summary Use identity-based comparison for IdentityHashMap#remove and #replace
+ */
+public class DefaultRemoveReplace {
+    public static void main(String[] args) {
+        final String key = "key";
+        final String internedValue = "value";
+        final String constructedValue = new String(new char[]{'v', 'a', 'l', 'u', 'e'});
+
+        if (!Objects.equals(internedValue, constructedValue)) {
+            throw new AssertionError("Bad test data: equality should hold");
+        }
+
+        if (internedValue == constructedValue) {
+            throw new AssertionError("Bad test data: identity should be different");
+        }
+
+        final IdentityHashMap<String, String> map = new IdentityHashMap<>();
+        map.put(key, internedValue);
+
+        if (map.replace(key, constructedValue, "dummy")) {
+            throw new AssertionError("Erroneously replacd value by equality");
+        }
+
+        if (!map.replace(key, internedValue, "dummy")) {
+            throw new AssertionError("Failed to replace value by identity");
+        }
+
+        map.clear();
+        map.put(key, constructedValue);
+
+        if (map.remove(key, internedValue)) {
+            throw new AssertionError("Erroneously removed value by equality");
+        }
+
+        if (!map.remove(key, constructedValue)) {
+            throw new AssertionError("Failed to remove value by identity");
+        }
+    }
+}
```

Comment:
Overall comments on this test. It does test the right set of four cases (positive and negative calls to `replace` and `remove`). However, it's **one** test that tests the four cases, instead of four tests. Using the same state makes it hard to add or maintain test cases in the future. It also trusts the return value of the method calls and doesn't make any assertions over the actual contents of the map. Without assertions over the expected contents, a method could behavior incorrectly and cause unrelated and confusing errors downstream, or even cause errors to be missed. I'd thus recommend the following:

* Convert this to a Test-NG test and use a `@BeforeTest` method to set up a test fixture consisting of a map containing the desired entries.
* Make each test case into its own test method that performs the method call and then makes assertions over the return value and expected result state of the map. Individual test methods is probably fine; no need to use a data provider for this.
* Probably add a "null hypothesis" test to ensure that the test fixture itself has the right properties, similar to the assertions at lines 38-44.
* Instead of fiddling around with strings, which have additional complexity caused by interning of string literals, I'd suggest using the technique I used in [JDK-8285295](https://bugs.openjdk.java.net/browse/JDK-8285295) and create a `record Box(int i) { }` class. With this it's easy to get multiple instances that are != but are equals.
* After further thought, maybe additional cases are necessary. For example, tests of calls to `remove` and `replace` with a key that is equals() but != to a key in the map.